include/deal.II/grid/grid_tools.h

```diff
@@ -1285,6 +1285,269 @@ namespace GridTools
+                              other_rank,
+                              request_index,
+                              found_intersection,
+                              requested_intersection);
+                          }
+                      }
+#else
+                    Assert(false, ExcNeedsCGAL());
+#endif
+                  }
+              }
+          }
+
+        if (answer.size() > 0)
+          send_ranks.emplace_back(other_rank);
+
+        return answer;
+      };
+
+      const auto process_answer = [&](const unsigned int other_rank,
+                                      const AnswerType & answer) {
+        for (unsigned int i = 0; i < answer.size(); ++i)
+          recv_components.emplace_back(other_rank,
+                                       answer[i].first,
+                                       answer[i].second);
+
+        if (answer.size() > 0)
+          recv_ranks.emplace_back(other_rank);
+      };
+
+      Utilities::MPI::ConsensusAlgorithms::selector<RequestType, AnswerType>(
+        potential_owners_ranks,
+        create_request,
+        answer_request,
+        process_answer,
+        comm);
+
+      // sort according to 1) intersection index and 2) rank (keeping the order
+      // of recv components with same indices and ranks)
+      std::stable_sort(recv_components.begin(),
+                       recv_components.end(),
+                       [&](const auto &a, const auto &b) {
+                         // intersecton index
+                         if (std::get<1>(a) != std::get<1>(b))
+                           return std::get<1>(a) < std::get<1>(b);
+
+                         // rank
+                         return std::get<0>(a) < std::get<0>(b);
+                       });
+
+      // sort according to 1) rank and 2) intersection index (keeping the
+      // order of recv components with same indices and ranks)
+      std::stable_sort(send_components.begin(),
+                       send_components.end(),
+                       [&](const auto &a, const auto &b) {
+                         // rank
+                         if (std::get<1>(a) != std::get<1>(b))
+                           return std::get<1>(a) < std::get<1>(b);
+
+                         // intersection idx
+                         return std::get<2>(a) < std::get<2>(b);
+                       });
+
+      // construct recv_ptrs
+      std::vector<unsigned int> n_found_intersections(
+        intersection_requests.size(), 0);
+      for (const auto &rc : recv_components)
+        ++n_found_intersections[std::get<1>(rc)];
+
+      recv_ptrs.resize(intersection_requests.size() + 1);
+      recv_ptrs[0] = 0;
+      for (unsigned int i = 0; i < n_found_intersections.size(); ++i)
+        recv_ptrs[i + 1] = recv_ptrs[i] + n_found_intersections[i];
+
+
+      // sort send and recv ranks
+      std::sort(recv_ranks.begin(), recv_ranks.end());
+      std::sort(send_ranks.begin(), send_ranks.end());
+
+
```

Comment:
```suggestion
```

Reply:
I am combining the PRs as suggested. This discussion is already ongoing in the other PR and therefore, I will not copy it over.